README.md
ecosystem-emulator-oop
======================

Experiments with OOP languages (by using a marine ecosystem)

This is mainly to illustrate differences between dynamic (with ruby) and static (with C++) OOP languages.


c++/Changelog
commit e6c40cb12a6296b6684777342fa017ae49ce52fd
Author: George Papanikolaou <g3orge.app@gmail.com>
Date:   Mon Jan 16 14:08:17 2012 +0200

    Rearranged classes and added age statistics

commit cda7de6cf478d6d2c636dccec08016ddf36a0769
Author: George Papanikolaou <g3orge.app@gmail.com>
Date:   Mon Jan 16 03:32:39 2012 +0200

    how i want it - dynamic

commit 42cc130b3d695ebe652ccec6cca71bb5a1246b8c
Author: George Papanikolaou <g3orge.app@gmail.com>
Date:   Mon Jan 16 03:08:31 2012 +0200

    Minor changes to the info option

commit 6a416de430b73af9ad5c82e41f891f950ec62f21
Author: George Papanikolaou <g3orge.app@gmail.com>
Date:   Mon Jan 16 02:52:34 2012 +0200

    Split into two files and improved eating proccess.

commit 1fbae798a77e8fb68dac898ef9bb035bc485423d
Author: George Papanikolaou <g3orge.app@gmail.com>
Date:   Sun Jan 15 23:11:30 2012 +0200

    Finishing touches. Changed the place of functions
    
    and changed some scopes.
    -Also added age code.

commit 2018763d4461718169ced34da6a1667a335df074
Author: George Papanikolaou <g3orge.app@gmail.com>
Date:   Sun Jan 15 22:36:02 2012 +0200

    Done with the eating/reproduction/movement
    
    -Added food and size stuff. When the food is 0, the size is reducing,
    and when the size is 0 they DIE.
    -Improved interaction and move.
    -The Plankton CANNOT die to preserve food in the system.
    -Also finished with the main step actions. Calling ALL the move().

commit ddb54abfd0ac5b471e5ecbf6077902b03aed9194
Author: George Papanikolaou <g3orge.app@gmail.com>
Date:   Sun Jan 15 21:19:06 2012 +0200

    Continuing the option. Done.
    
    -finished with the Info option.
        Added counters (specific) and print them along with general info
    -finished with the add option.
        Improved the adding process,
        added the "Infection" code
    
    Maybe there is THE NEED FOR >>*<<map[i][j] where I access the cells

commit dcf383a376ac9f594c34480dfe6d33a79961683a
Author: George Papanikolaou <g3orge.app@gmail.com>
Date:   Sun Jan 15 18:28:58 2012 +0200

    Improving menus and classes
    
    -Improving The Add option in the menu
    -Implementing the interaction (for eating and reproduction)
    -Implementing the move for all the cells
    -Adding Shark and Dolphin in the create function so it can be used in
        the adding and the reproduction options...
    Now I have to make the actual movement

commit a2cbd4927f660a17bf71ffb205d2bfa6bb7b6da2
Author: George Papanikolaou <g3orge.app@gmail.com>
Date:   Sun Jan 15 16:20:16 2012 +0200

    Implementing the creator and the interaction
    
    -Also, the menus are all done, except the main "step" and the info,
    in which I need to add more specific stuff
    -Done with the creator function
    -Starting to implement the eating the the interaction

commit b6875f168222ec8fb5877d43f5ffca13f4aea854
Author: George Papanikolaou <g3orge.app@gmail.com>
Date:   Sat Jan 14 21:00:39 2012 +0200

    I'm in the middle of creating the function of GOD
    
    Actually it is the which_Organism function to create organism objects
    Also continued with the move and the interaction functions, and
    created the NON-ORGANISM-OBJECT

commit a8db245a8bfffa6620ea0872070ee87b8543739f
Author: George Papanikolaou <g3orge.app@gmail.com>
Date:   Sat Jan 14 17:58:50 2012 +0200

    Getting started with the advanced map generation
    
    Using vectors to generate map - step 1 made.
    Also created labels for the goto-s of the info/print loop

commit d50b68d8a912b0c533d18553bd0ba62c33ece605
Author: George Papanikolaou <g3orge.app@gmail.com>
Date:   Sat Jan 14 17:22:58 2012 +0200

    Update README

commit 8f632bc4e34be411a38a5392b59ec6d40b4e1120
Author: George Papanikolaou <g3orge.app@gmail.com>
Date:   Sat Jan 14 17:16:32 2012 +0200

    new README

commit 23bdf198f1a2c641a948fb1b3b37c5b21ecfdd2d
Author: George Papanikolaou <g3orge.app@gmail.com>
Date:   Sat Jan 14 17:14:06 2012 +0200

    new README

commit 53e8515d1f3bcb0bd2a8d24a9bf754a4d8efa0f2
Author: George Papanikolaou <g3orge.app@gmail.com>
Date:   Sat Jan 14 17:07:33 2012 +0200

    Get started with more stuff inside the sub-classes
    
    I just created the age, the size and the alias for all the basic
    organisms and improved the whole class system
    TODO: Create the map - Probaly with vector and pointers

commit b08419469199a43b85ddfca4850c5300712e0887
Author: George Papanikolaou <g3orge.app@gmail.com>
Date:   Fri Jan 13 17:00:54 2012 +0200

    Implementing things into main and creating classes
    
    I just created the prototypes of all the required classes,
    and started implementing more things inside the main, the menus,
    and the map generation code.
    TODO: Add the REAL stuff into the classes
          put some organisms into the map.

commit 618af021db8126ecc0767d33538a14552c25105c
Author: George Papanikolaou <g3orge.app@gmail.com>
Date:   Fri Jan 13 02:19:58 2012 +0200

    This is a whole new implementation. Basically, v2

commit d8e31562dbc83ea387b5617e48e53dfbea30c225
Author: George Papanikolaou <g3orge.app@gmail.com>
Date:   Sun Jan 8 17:43:41 2012 +0200

    changes in random and wait

commit 469bc4a6cae4ae0db7ec9e1a89d82f986868c412
Author: George Papanikolaou <g3orge.app@gmail.com>
Date:   Thu Dec 15 21:12:40 2011 +0200

    More

commit cfdd0861e2f9a3e04b644d754859814a0c805245
Author: George Papanikolaou <g3orge.app@gmail.com>
Date:   Thu Dec 15 19:05:46 2011 +0200

    Getting started with time stuff

commit b5ae79707873c5798008b70617bb8117126a8999
Author: George Papanikolaou <g3orge.app@gmail.com>
Date:   Thu Dec 15 15:46:44 2011 +0200

    Creating the map and placing organisms

commit 0b0ef151efd2d85e4046f0a005e21fef4e1e2c93
Author: George Papanikolaou <g3orge.app@gmail.com>
Date:   Thu Dec 15 14:33:47 2011 +0200

    Starting to create the main function

commit e2d4325f46f51613e4b41f8c4dbe78fbd1dc389a
Author: George Papanikolaou <g3orge.app@gmail.com>

ruby/eco.rb
#
# Ecosystem emulator in ruby
# George 'papanikge' Papanikolaou
# This is mainly to experiment with the dynamic OOP languages
#

require './organisms.rb'

#
# Our imaginary marine world (with metadata et all)
#
class World
  include Enumerable

  attr_reader :dim_x
  attr_reader :dim_y
  attr_reader :org_count
  # this is the actual array that keeps the organisms
  # TODO: should we make the 'World' class a subclass of Array?
  attr_accessor :map

  def initialize
    @org_count = 0
    # possible ranges are hard coded because I'm bored
    @dim_x = rand(15..30)
    @dim_y = rand(15..30)
    init_world(@dim_x,@dim_y)
  end

  #
  # Runs a block of code against our world (needs a block)
  #
  def traverse_map
    @map.each_with_index do |xs, i|
      xs.each_with_index do |ys, j|
        @map[i][j] = yield(ys)
      end
    end
  end

  #
  # Create or reinitialize a 'world' array for the game
  #
  def init_world(x, y)
    @map = Array.new(x) { |c| c = Array.new(y) }
    @map.each_with_index do |xs, i|
      xs.each_with_index do |ys, j|
        @map[i][j] = gen_random_org(i, y)
        @org_count += 1
      end
    end
  end

  #
  # One unit of time (for the world)
  #
  def step
    traverse_map do |el|
      fail "cell does not contain organism" unless el.class.ancestors.include? Organism
      el.tick
    end
  end

  #
  # Add an organism to a specific cell. This also updates metadata
  #
  def add(org, x, y)
    puts "You requested a #{org.class} in #{x},#{y}."
    if @map[x][y]
      # This may be an example of bad programming. We cannot use #eat_or_interact
      # here because it only handles organisms inside our world.
      if @map[x][y] > org.size
        @map[x][y].orgs_eaten += org.size
        org = nil
      else
        org.orgs_eaten += @map[x][y].size
        org.x = x
        org.y = y
        @map[x][y] = org
      end
    else
      # there was nothing there
      @map[x][y] = org
    end
    puts "Added"
    @org_count += 1
  end

  #
  # Adds an infection to a cell of the map that kills all the neighbor cells.
  # Needs a level of infection to determine how many neighbors to kill.
  # Works recursively.
  #
  def infect(level, x, y)
    return unless level > 0
    (1..9).each do |n|
      coords = @map[x][y].get_coords(n)
      infect(level - 1, coords[0],coords[1])  # Recursive step
      @map[coords[0]][coords[1]] = nil        # Die...
    end
  end

  #
  # Prints info (of course)
  #
  def info
    # TODO: add Graphics (see OpenGL or Shoes)
    puts "The map is: #{@dim_x}x#{@dim_y}, and contains #{@org_count} organisms."
  end

  #
  # Functions responsible for getting user input, and acting accordingly
  #

  def get_user_coords
    x ||= rand(0..@dim_x)   # fail-safes. Also we probably need these
    y ||= rand(0..@dim_y)   # because ruby scopes...
    puts "Reminder that the map is: #{@dim_x}x#{@dim_y}"
    loop do
      print "x? "
      x = STDIN.gets.chomp.to_i
      break if (0..@dim_x).include? x
    end
    loop do
      print "y? "
      y = STDIN.gets.chomp.to_i
      break if (0..@dim_y).include? y
    end
    return [x,y]
  end

  #
  # Wrapper aroung World#add
  #
  def user_add
    choice ||= gen_random_org
    puts "Available organisms to add:"
    $orgs.each { |org| puts org }
    input = STDIN.gets.chomp
    input.capitalize!
    $orgs.each do |org|
      if org.include? input
        choice = eval "#{org}.new"
        break
      end
    end
    c = get_user_coords
    choice.x = c[0]
    choice.y = c[1]
    add(choice, c[0], c[1])
  end

  def user_infect
    level ||= rand(1..4)
    puts "Level of infection? "
    loop do
      level = STDIN.gets.chomp.to_i
      break if level < 30
      puts "Too big. You're mean. "
    end
    c = get_user_coords
    infect(level, c[0], c[1])
  end
end

#
# Main thread of execution
#

puts "Welcome to this (hardly) funny little game. :)"
puts "Generating the map..."
$world = World.new
puts "Done. The ecosystem map will have #{$world.dim_x}x#{$world.dim_y} dimensions."
print "Options:\tS: step\n\tA: add\n\tV: infect\n\tI: info\n\tR: restart\n\tQ: quit\n"
loop do
  puts ">>> Main menu <<<"
  input = STDIN.gets.chomp
  input.downcase!
  case input
  when 's' then $world.step
  when 'a' then $world.user_add
  when 'v' then $world.user_infect
  when 'i' then $world.info
  when 'r' then $world.init_world
  when 'q' then break
  else puts "Wrong option."
  end
end


src/eco.cpp
// This is the first file of the C++ Project 'EcoSystem', that I'm working on.
// This is the compiled file.
// © George Papanikolaou 2011

#include <iostream>
#include <cstring>
#include <ctime>
#include <vector>
#include <cstdlib>
#include "classes.h"
#include "functions.h"

using namespace std;

int RandomInteger(int low, int high); // Declaration
Organism create_Organism (int digit);

int main() {
    restart_label:      // This is the required label for the restart action
    srand((int)time(0));
    NON E;              // Creating the empty object
    int x=RandomInteger(15,20);
    int y=RandomInteger(15,20);
    cout << "The map will have " << x << " length and" << y << " width" << endl;
    int N=x*y;
    cout << "Generating the system!" << endl;
    wait(1);
    Organism map[x][y];   // Trying to create a container for all the objects - The map
    // Counters:
    int step_counter = 1;
    int org_counter = 0;
    int dead_org = 0;
    cout << "Placing organisms into the map..." << endl;
    for (int i = 0; i <= x; i++) {
        for (int j = 0; j <= y; j++) {
                int digit = RandomInteger(0,20);    // Generate a random digit for the organism creation 
                Organism Y = create_Organism (digit);
                Y.where_x=i;              // Add their place inside the placeholders
                Y.where_y=j;
                map[i][j] = Y;
                if ( digit >= 4 ) org_counter++;   // Add one to the organism counter if an organism is been added
        }
    }
    // Placing a shark and a dolphin.
    bool flag=false;
    while (flag==false) {
        int place_x = RandomInteger(0,N);
        int place_y = RandomInteger(0,N);
        if (map[place_x][place_y].alias='O') {
            Shark S;
            map[place_x][place_y] = S;
            flag=true;
            org_counter++;
        }
    }
    flag=false;
    while (flag==false) {
        int place_x = RandomInteger(0,N);
        int place_y = RandomInteger(0,N);
        if (map[place_x][place_y].alias='O') {
            Dolphin D;
            map[place_x][place_y] = D;
            flag=true;
            org_counter++;
        }
    }
    cout << "Map initialized successfully" << endl;
    // Done. Starting menus and steps.
    char choice;
    info_print_label:      // <-------- This is the required label for the print and the info actions
    cout << "Press S to make a step\nP to print the map\nA to add an Organism or Infection\nR to restart\nI to get info\nAnything else to quit" << endl;
    cin >> choice;
    while (choice == 'S' || choice == 's') {
        for (int i=0;i=x;i++) {
            for (int j=0; j=y; j++) {
                map[i][j].move();           // Calling the move function of all the organisms to make a step  
            }
        }
        step_counter++;
        cout << "Press S to make a step\nP to print the map\nA to add an organism or Infection\nR to restart\nI to get info\nAnything else to quit" << endl;
        // Yes you have need to hit 's' everytime, for every step. I need something like conio.h for *nix to implement a continuous loop.
        cin >> choice;
    }
    if (choice == 'R' || choice == 'r')
        goto restart_label;
    else if (choice == 'I' || choice == 'i') {
        // Starting to count organisms:
        int P_cn=0, Z_cn=0, K_cn=0, M_cn=0, G_cn=0, T_cn=0, C_cn=0, A_cn=0, D_cn=0, S_cn=0;    // Counters
        int P_age=0, Z_age=0, K_age=0, M_age=0, G_age=0, T_age=0, C_age=0, A_age=0, D_age=0, S_age=0;    // Total age counters
        for (int i=0; i=x; i++) {
            for (int j=0; j=y; j++) {
                // Maybe there is THE NEED FOR >>*<<map[i][j] down below here
                if ( map[i][j].alias == 'P' ) {
                    P_cn++;
                    P_age = P_age+map[i][j].age;        // Creating total organism age counters for average.
                }
                else if ( map[i][j].alias == 'Z' ) {
                    Z_cn++;
                    Z_age = Z_age+map[i][j].age;
                }
                else if ( map[i][j].alias == 'K' ) {
                    K_cn++;
                    K_age = K_age+map[i][j].age;
                }
                else if ( map[i][j].alias == 'M' ) {
                    M_cn++;
                    M_age = M_age+map[i][j].age;
                }
                else if ( map[i][j].alias == 'G' ) {
                    G_cn++;
                    G_age = G_age+map[i][j].age;
                }
                else if ( map[i][j].alias == 'T' ) {
                    T_cn++;
                    T_age = T_age+map[i][j].age;
                }
                else if ( map[i][j].alias == 'C' ) {
                    C_cn++;
                    C_age = C_age+map[i][j].age;
                }
                else if ( map[i][j].alias == 'A' ) {
                    A_cn++;
                    A_age = A_age+map[i][j].age;
                }
                else if ( map[i][j].alias == 'S' ) {
                    S_cn++;
                    S_age = S_age+map[i][j].age;
                }
                else if ( map[i][j].alias == 'D' ) {
                    D_cn++;
                    D_age = D_age+map[i][j].age;
                }
        }
        cout << "There are " << org_counter << " total organisms in the map." << endl;
        cout << "The spesific organisms are:\n Phytoplankton(P) - " << P_cn << "organisms with " << P_age/P_cn << " age average."
             << "\n Zooplankton(Z) - " << Z_cn << "organisms with " << Z_age/Z_cn << " age average."
             << "\n Squid(K) - "<< K_cn << "organisms with " << K_age/K_cn << " age average."
             << "\n Mylittus(M) - " << M_cn << "organisms with " << M_age/M_cn << " age average."
             << "\n Shrimp(G) - " << G_cn << "organisms with " << G_age/G_cn << " age average."
             << "\n Octopus(T) - "<< T_cn << "organisms with " << T_age/T_cn << " age average."
             << "\n Crab(C) - " << C_cn << "organisms with " << C_age/C_cn << " age average."
             << "\n Sprattus(A) - " << A_cn << "organisms with " << A_age/A_cn << " age average."
             << "\n Dolphin(D) - " << D_cn << "organisms with " << D_age/D_cn << " age average."
             << "\n Shark(S) - " << S_cn << "organisms with " << S_age/S_cn << " age average." << endl;
        cout << "\nTotal dead organisms: " << dead_org << endl;
        cout << "System running for " << clock()/CLOCKS_PER_SEC << " seconds" << endl;
        goto info_print_label;
    }
     else if (choice == 'A' || choice == 'a') {
        // User organism creation loop:
        cout << "Press the alias of the Organism you want to add or I to add an infection" << endl;
        char adding;
        cin >> adding;
        if ( adding == 'I' || adding == 'i' ) {
            // The Infection code here:
            cout << "Press the number of the row you want the infection to appear." << endl;
            int inf_x;
            cin >> inf_x;
            cout << "Press the number of the collumn." << endl;
            int inf_y;
            cin >> inf_y;
            // Killing the near organisms:
            if ( inf_x == 0 ) {
                if (inf_y == 0 || inf_y == y) {
                    map[inf_x][inf_y] = E;     // Give the pointer to that empty object
                    map[inf_x+1][inf_y] = E;   // For the near cells ( without getting out of limits)
                    map[inf_x][inf_y+1] = E;
                    map[inf_x+1][inf_y+1] = E;
                }
                else {
                    map[inf_x][inf_y] = E;
                    map[inf_x+1][inf_y] = E;
                    map[inf_x][inf_y+1] = E;
                    map[inf_x+1][inf_y+1] = E;
                    map[inf_x][inf_y-1] = E;
                    map[inf_x+1][inf_y-1] = E;
                }
            }
            else if (inf_y == 0 || inf_y == y) {
                map[inf_x][inf_y] = E;
                map[inf_x+1][inf_y] = E;
                map[inf_x][inf_y+1] = E;
                map[inf_x+1][inf_y+1] = E;
                map[inf_x-1][inf_y+1] = E;
                map[inf_x-1][inf_y] =E;
            }
            else if ( inf_x == x ) {
                if (inf_y == 0 || inf_y == y) {
                    map[inf_x][inf_y+1] = E;
                    map[inf_x-1][inf_y+1] = E;
                    map[inf_x-1][inf_y] = E;
                }
                else {
                    map[inf_x][inf_y+1] = E;
                    map[inf_x-1][inf_y+1] = E;
                    map[inf_x-1][inf_y] = E;
                    map[inf_x-1][inf_y-1] = E;
                }
            else {
                map[inf_x][inf_y] = E;
                map[inf_x+1][inf_y] = E;
                map[inf_x][inf_y+1] = E;
                map[inf_x-1][inf_y] = E;
                map[inf_x][inf_y-1] = E;
                map[inf_x+1][inf_y+1] = E;
                map[inf_x-1][inf_y-1] = E;
                map[inf_x+1][inf_y-1] = E;
                map[inf_x-1][inf_y+1] = E;
            }
        }
        else {
            cout << "Press the number of the row you want the Organism to be added" << endl;
            int arow;
            cin >> arow;
            cout << "Now press the number of the collumn you want the Organism to be added" << endl;
            int acollumn;
            cin >> acollumn;
            int c;
            if ( adding == 'P' || adding == 'p' ) c=4;         // Matching aliases with codes for the create_Organism function
            else if ( adding == 'Z' || adding == 'z' ) c=8;
            else if ( adding == 'K' || adding == 'k' ) c=11;
            else if ( adding == 'M' || adding == 'm' ) c=13;
            else if ( adding == 'G' || adding == 'g' ) c=15;
            else if ( adding == 'T' || adding == 't' ) c=18;
            else if ( adding == 'C' || adding == 'c' ) c=19;
            else if ( adding == 'A' || adding == 'a' ) c=20;
            else if ( adding == 'D' || adding == 'd' ) c=21;
            else if ( adding == 'S' || adding == 's' ) c=22;
            else { 
                cout << "There is no " << adding << " alias in the system" << endl;
                goto info_print_label;
            }
            Organism Q = create_Organism (c);
            map[arow][acollumn] = Q;
            Q.where_x = arow;
            Q.where_y = acollumn;
            cout << "Success" << endl;
            goto info_print_label;
        }
    }
    else if (choice == 'P' || choice == 'p') {
        // Print option: (for the console)
        int i,j;
        for (i=0; i=x; i++) {
            for (j=0; j=y; j++) {
                // Maybe there is THE NEED FOR >>*<<map[i][j] down below here
                cout << map[i][j].alias << "-";
            }
            cout << "\b" << endl;   //Change the line in the stdout
        }
        cout << "-------------------------------";
        goto info_print_label;
    }
    else { cout << "System quitting after " << clock()/CLOCKS_PER_SEC << " seconds" << endl; }
    return 0;
}


c++/src/eco.cc
//===========================================
// C++ simple Ecosystem emulator
// © George Papanikolaou 2011-2013
// See LICENSE
//===========================================

#include <iostream>
#include <cstring>
#include <ctime>
#include <vector>
#include <cstdlib>
#include "classes.h"

using namespace std;

// Generate a random digit between arguments
int random(int low, int high)
{  
    int diff = high - low + 1;

    return(low + rand() % diff);
} 

// Generate a random organism
Organism gen_random_org(int digit)
{
    NON B;
    if (digit==0) { return B; }   // just a abstract pointer for non-animal cell
    else if (digit==1) { return B; } 
    else if (digit==2) { return B; } 
    else if (digit==3) { return B; } 
    else if (digit==4) { Phytoplankton X; return X; } /*****************************/
    else if (digit==5) { Phytoplankton X; return X; } /*   Phytoplankton creation  */
    else if (digit==6) { Phytoplankton X; return X; } /* The same times as "empty" */
    else if (digit==7) { Phytoplankton X; return X; } /*****************************/
    else if (digit==8) { Zooplankton X; return X; } 
    else if (digit==9) { Zooplankton X; return X; }
    else if (digit==10) { Zooplankton X; return X; }
    else if (digit==11) { Squid X; return X; } 
    else if (digit==12) { Squid X; return X; }
    else if (digit==13) { Mylittus X; return X; } 
    else if (digit==14) { Mylittus X; return X; } 
    else if (digit==15) { Shrimp X; return X; } 
    else if (digit==16) { Shrimp X; return X; } 
    else if (digit==17) { Shrimp X; return X; } 
    else if (digit==18) { Octopus X; return X; } 
    else if (digit==19) { Crab X; return X; } 
    else if (digit==20) { Sprattus X; return X; } 
    else if (digit==21) { Dolphin X; return X; }    // These are for the reproduction and the add option. NO RANDOM.
    else if (digit==22) { Shark X; return X; } 
}

int main()
{
restart_label:      // This is the required label for the restart action
    srand((int)time(0));
    NON E;              // Creating the empty object
    int x=RandomInteger(15,20);
    int y=RandomInteger(15,20);
    cout << "The map will have " << x << " length and" << y << " width" << endl;
    int N=x*y;
    cout << "Generating the system!" << endl;
    Organism map[x][y];   // Trying to create a container for all the objects - The map
    // Counters:
    int step_counter = 1;
    int org_counter = 0;
    int dead_org = 0;
    cout << "Placing organisms into the map..." << endl;
    for (int i = 0; i <= x; i++) {
        for (int j = 0; j <= y; j++) {
            int digit = RandomInteger(0,20);    // Generate a random digit for the organism creation 
            Organism Y = create_Organism (digit);
            Y.where_x=i;              // Add their place inside the placeholders
            Y.where_y=j;
            map[i][j] = Y;
            if ( digit >= 4 ) org_counter++;   // Add one to the organism counter if an organism is been added
        }
    }
    // Placing a shark and a dolphin.
    bool flag=false;
    while (flag==false) {
        int place_x = RandomInteger(0,N);
        int place_y = RandomInteger(0,N);
        if (map[place_x][place_y].alias='O') {
            Shark S;
            map[place_x][place_y] = S;
            flag=true;
            org_counter++;
        }
    }
    flag=false;
    while (flag==false) {
        int place_x = RandomInteger(0,N);
        int place_y = RandomInteger(0,N);
        if (map[place_x][place_y].alias='O') {
            Dolphin D;
            map[place_x][place_y] = D;
            flag=true;
            org_counter++;
        }
    }
    cout << "Map initialized successfully" << endl;
    // Done. Starting menus and steps.
    char choice;
info_print_label:      // <-------- This is the required label for the print and the info actions
    cout << "Press S to make a step\nP to print the map\nA to add an Organism or Infection\nR to restart\nI to get info\nAnything else to quit" << endl;
    cin >> choice;
    while (choice == 'S' || choice == 's') {
        for (int i=0;i=x;i++) {
            for (int j=0; j=y; j++) {
                map[i][j].move();           // Calling the move function of all the organisms to make a step  
            }
        }
        step_counter++;
        cout << "Press S to make a step\nP to print the map\nA to add an organism or Infection\nR to restart\nI to get info\nAnything else to quit" << endl;
        // Yes you have need to hit 's' everytime, for every step. I need something like conio.h for *nix to implement a continuous loop.
        cin >> choice;
    }
    if (choice == 'R' || choice == 'r')
        goto restart_label;
    else if (choice == 'I' || choice == 'i') {
        // Starting to count organisms:
        int P_cn=0, Z_cn=0, K_cn=0, M_cn=0, G_cn=0, T_cn=0, C_cn=0, A_cn=0, D_cn=0, S_cn=0;    // Counters
        int P_age=0, Z_age=0, K_age=0, M_age=0, G_age=0, T_age=0, C_age=0, A_age=0, D_age=0, S_age=0;    // Total age counters
        for (int i=0; i=x; i++) {
            for (int j=0; j=y; j++) {
                // Maybe there is THE NEED FOR >>*<<map[i][j] down below here
                if ( map[i][j].alias == 'P' ) {
                    P_cn++;
                    P_age = P_age+map[i][j].age;        // Creating total organism age counters for average.
                }
                else if ( map[i][j].alias == 'Z' ) {
                    Z_cn++;
                    Z_age = Z_age+map[i][j].age;
                }
                else if ( map[i][j].alias == 'K' ) {
                    K_cn++;
                    K_age = K_age+map[i][j].age;
                }
                else if ( map[i][j].alias == 'M' ) {
                    M_cn++;
                    M_age = M_age+map[i][j].age;
                }
                else if ( map[i][j].alias == 'G' ) {
                    G_cn++;
                    G_age = G_age+map[i][j].age;
                }
                else if ( map[i][j].alias == 'T' ) {
                    T_cn++;
                    T_age = T_age+map[i][j].age;
                }
                else if ( map[i][j].alias == 'C' ) {
                    C_cn++;
                    C_age = C_age+map[i][j].age;
                }
                else if ( map[i][j].alias == 'A' ) {
                    A_cn++;
                    A_age = A_age+map[i][j].age;
                }
                else if ( map[i][j].alias == 'S' ) {
                    S_cn++;
                    S_age = S_age+map[i][j].age;
                }
                else if ( map[i][j].alias == 'D' ) {
                    D_cn++;
                    D_age = D_age+map[i][j].age;
                }
            }
            cout << "There are " << org_counter << " total organisms in the map." << endl;
            cout << "The spesific organisms are:\n Phytoplankton(P) - " << P_cn << "organisms with " << P_age/P_cn << " age average."
                << "\n Zooplankton(Z) - " << Z_cn << "organisms with " << Z_age/Z_cn << " age average."
                << "\n Squid(K) - "<< K_cn << "organisms with " << K_age/K_cn << " age average."
                << "\n Mylittus(M) - " << M_cn << "organisms with " << M_age/M_cn << " age average."
                << "\n Shrimp(G) - " << G_cn << "organisms with " << G_age/G_cn << " age average."
                << "\n Octopus(T) - "<< T_cn << "organisms with " << T_age/T_cn << " age average."
                << "\n Crab(C) - " << C_cn << "organisms with " << C_age/C_cn << " age average."
                << "\n Sprattus(A) - " << A_cn << "organisms with " << A_age/A_cn << " age average."
                << "\n Dolphin(D) - " << D_cn << "organisms with " << D_age/D_cn << " age average."
                << "\n Shark(S) - " << S_cn << "organisms with " << S_age/S_cn << " age average." << endl;
            cout << "\nTotal dead organisms: " << dead_org << endl;
            cout << "System running for " << clock()/CLOCKS_PER_SEC << " seconds" << endl;
            goto info_print_label;
        }
        else if (choice == 'A' || choice == 'a') {
            // User organism creation loop:
            cout << "Press the alias of the Organism you want to add or I to add an infection" << endl;
            char adding;
            cin >> adding;
            if ( adding == 'I' || adding == 'i' ) {
                // The Infection code here:
                cout << "Press the number of the row you want the infection to appear." << endl;
                int inf_x;
                cin >> inf_x;
                cout << "Press the number of the collumn." << endl;
                int inf_y;
                cin >> inf_y;
                // Killing the near organisms:
                if ( inf_x == 0 ) {
                    if (inf_y == 0 || inf_y == y) {
                        map[inf_x][inf_y] = E;     // Give the pointer to that empty object
                        map[inf_x+1][inf_y] = E;   // For the near cells ( without getting out of limits)
                        map[inf_x][inf_y+1] = E;
                        map[inf_x+1][inf_y+1] = E;
                    }
                    else {
                        map[inf_x][inf_y] = E;
                        map[inf_x+1][inf_y] = E;
                        map[inf_x][inf_y+1] = E;
                        map[inf_x+1][inf_y+1] = E;
                        map[inf_x][inf_y-1] = E;
                        map[inf_x+1][inf_y-1] = E;
                    }
                }
                else if (inf_y == 0 || inf_y == y) {
                    map[inf_x][inf_y] = E;
                    map[inf_x+1][inf_y] = E;
                    map[inf_x][inf_y+1] = E;
                    map[inf_x+1][inf_y+1] = E;
                    map[inf_x-1][inf_y+1] = E;
                    map[inf_x-1][inf_y] =E;
                }
                else if ( inf_x == x ) {
                    if (inf_y == 0 || inf_y == y) {
                        map[inf_x][inf_y+1] = E;
                        map[inf_x-1][inf_y+1] = E;
                        map[inf_x-1][inf_y] = E;
                    }
                    else {
                        map[inf_x][inf_y+1] = E;
                        map[inf_x-1][inf_y+1] = E;
                        map[inf_x-1][inf_y] = E;
                        map[inf_x-1][inf_y-1] = E;
                    }
                    else {
                        map[inf_x][inf_y] = E;
                        map[inf_x+1][inf_y] = E;
                        map[inf_x][inf_y+1] = E;
                        map[inf_x-1][inf_y] = E;
                        map[inf_x][inf_y-1] = E;
                        map[inf_x+1][inf_y+1] = E;
                        map[inf_x-1][inf_y-1] = E;
                        map[inf_x+1][inf_y-1] = E;
                        map[inf_x-1][inf_y+1] = E;
                    }
                }
                else {
                    cout << "Press the number of the row you want the Organism to be added" << endl;
                    int arow;
                    cin >> arow;
                    cout << "Now press the number of the collumn you want the Organism to be added" << endl;
                    int acollumn;
                    cin >> acollumn;
                    int c;
                    if ( adding == 'P' || adding == 'p' ) c=4;         // Matching aliases with codes for the create_Organism function
                    else if ( adding == 'Z' || adding == 'z' ) c=8;
                    else if ( adding == 'K' || adding == 'k' ) c=11;
                    else if ( adding == 'M' || adding == 'm' ) c=13;
                    else if ( adding == 'G' || adding == 'g' ) c=15;
                    else if ( adding == 'T' || adding == 't' ) c=18;
                    else if ( adding == 'C' || adding == 'c' ) c=19;
                    else if ( adding == 'A' || adding == 'a' ) c=20;
                    else if ( adding == 'D' || adding == 'd' ) c=21;
                    else if ( adding == 'S' || adding == 's' ) c=22;
                    else { 
                        cout << "There is no " << adding << " alias in the system" << endl;
                        goto info_print_label;
                    }
                    Organism Q = create_Organism (c);
                    map[arow][acollumn] = Q;
                    Q.where_x = arow;
                    Q.where_y = acollumn;
                    cout << "Success" << endl;
                    goto info_print_label;
                }
            }
            else if (choice == 'P' || choice == 'p') {
                // Print option: (for the console)
                int i,j;
                for (i=0; i=x; i++) {
                    for (j=0; j=y; j++) {
                        // Maybe there is THE NEED FOR >>*<<map[i][j] down below here
                        cout << map[i][j].alias << "-";
                    }
                    cout << "\b" << endl;   //Change the line in the stdout
                }
                cout << "-------------------------------";
                goto info_print_label;
            }
            else { cout << "System quitting after " << clock()/CLOCKS_PER_SEC << " seconds" << endl; }
            return 0;
        }
    }
}
